sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala

```diff
@@ -203,6 +203,10 @@ case class AttributeReference(
     case _ => false
   }
 
+  override def semanticHash(): Int = {
+   this.exprId.hashCode()
```

Comment:
It works fine currently because by the time this has run, the expression is a bound attribute. We should consider running the logic of finding subexpressions earlier, as chenghao suggests. I added a unit test that exposes the problem of not defining this. As a general rule, I think any class that defines semanticEquals should define semanticHash, even if it is the default implementation.